File: leetcode/Algorithms/501.FindModeInBinarySearchTree/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int[] findMode(TreeNode root) {
        HashMap<Integer, Integer> map = new HashMap<>();
        travel(root, map);

        ArrayList<Integer> tmp = new ArrayList<>();
        int max = 0;
        for (Map.Entry<Integer, Integer> entry : map.entrySet()) {
            if (entry.getValue() > max) {
                tmp.clear();
                tmp.add(entry.getKey());
                max = entry.getValue();
            } else if (entry.getValue() == max) {
                tmp.add(entry.getKey());
            }
        }

        int[] result = new int[tmp.size()];
        for (int i = 0; i < tmp.size(); i++) result[i] = tmp.get(i);
        return result;
    }

    private void travel(TreeNode treeNode, HashMap<Integer, Integer> map) {
        if (treeNode != null) {
            map.put(treeNode.val, map.getOrDefault(treeNode.val, 0) + 1);
            if (treeNode.left != null) travel(treeNode.left, map);
            if (treeNode.right != null) travel(treeNode.right, map);
        }
    }
}

File: leetcode/Algorithms/501.FindModeInBinarySearchTree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.current_value = 0
        self.current_count = 0
        self.max_count = 0
        self.result = []

    def findMode(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        self.morris_traversal(root)
        return self.result

    def update(self, val):
        if val != self.current_value:
            self.current_count = 1
        else:
            self.current_count += 1

        self.current_value = val

        if self.current_count > self.max_count:
            self.max_count = self.current_count
            self.result = [self.current_value]
        elif self.current_count == self.max_count:
            self.result.append(self.current_value)

    def morris_traversal(self, root):
        node = root

        while node is not None:
            if node.left is None:
                self.update(node.val)
                node = node.right
            else:
                pre = node.left
                while (pre.right is not None) and (pre.right != node):
                    pre = pre.right

                if pre.right is None:
                    pre.right = node
                    node = node.left
                else:
                    pre.right = None
                    self.update(node.val)
                    node = node.right

